### PR TITLE
Gateway: pass through errors encountered when running final execute

### PR DESCRIPTION
**Pull request by princed**
Fixes a problem where `@apollo/gateway` hides errors encountered when running final `execute()` from `graphql-js`, which can potentially both throw (e.g. from [here](https://github.com/graphql/graphql-js/blob/fbe8402db3b7751c5b1340886c3b09a7f36aee58/src/execution/execute.js#L375)) *and* return `errors` array in following cases:

- https://github.com/graphql/graphql-js/blob/fbe8402db3b7751c5b1340886c3b09a7f36aee58/src/execution/execute.js#L401-L404 
- https://github.com/graphql/graphql-js/blob/fbe8402db3b7751c5b1340886c3b09a7f36aee58/src/execution/execute.js#L237-L239
---
Originally at: apollographql/apollo-server#4523

### CODE REVIEW

**Comment by princed** [comment]
@abernix Any chance to get an actual review? 😅

**Comment by princed** [comment]
@abernix Does this have any chance to be merged? Or shall close it and just wait until the rewrite lands?

**Comment by malyusha** [comment]
TL;DR
**Merge it, please.**

I recently encountered a bug, when data of field received from federation gateway was `null`, but the service responsible for that data returned correct array of objects. After spending about 5hrs I finally got to execute context from `graphql-js` and logged errors that were created after `parseField` method call. The bug was that underlying service is written in Go, using `gqlgen` library. All `Int` types are casted to `int64`, but JS graphql raises error for numbers overflowing `int32`. I was not able to see any error logs from federation even in console.

**Comment by glasser** [comment]
While this seems like a problem worth solving, I'm curious if we should also null out `data` in the case that there are errors.

**Comment by glasser** [comment]
@malyusha Also FYI it sounds like this is a bug in gqlgen — [Ints are 32 bits in GraphQL](https://github.com/apollographql/apollo-server/pull/5039) and

> If the integer internal value represents a value less than -2^31 or greater than or equal to 2^31, a field error should be raised

**Comment by malyusha** [comment]
@glasser yes, this is more about not following graphql spec on `gqlgen` side. But our Go service with graphql has already spread monolithic graphql schema (which we're currently splitting using federation), and we can not afford force update for mobile clients. So one of object fields has `Int` type, and we must live with the fact, that the value under this `Int`, in some cases, is greater than possible 2147483647. 
By now, I've fixed it with tmp hack, replacing `2147483647` and `-2147483648` inside `node_modules/graphql/types` with `Number.MAX_SAFE_INTEGER` & `Number.MIN_SAFE_INTEGER` respectively..

Personally, as a developer, I'd like to decide whether I need to run `execute` or not after schema being received from federated services. In some cases, I think, it's not worth it. Correct me if i'm wrong.

**Comment by glasser** [comment]
The main point of this run is to strip out fields that haven't been requested but that are part of the response for internal reasons (like fields from `@requires`). There certainly other ways imaginable to implement it though.

**Comment by princed** [comment]
Same story, I spent ages figuring out why the data returned from my gateway was silently null while it was present in response of my Go service. 

@malyusha In case you want to avoid the issue now and skip waiting until Apollo maintainers decide what to do for another half a year, I have published a fork with this PR - https://www.github.com/princed/apollo-server/tree/forked-package/packages%2Fapollo-gateway%2Fpackage.json
I haven't updated it ever since though.

**Review by martijnwalraven — APPROVED** [review]
@princed Thanks for investigating and fixing this issue. I'm sorry this took so long to merge.

**Comment by princed** [comment]
Thank you @martijnwalraven!

**Comment by glasser** [comment]
I think there's a bug here. Let's say we have a query `{ x }` where `x` is a non-nullable field, and the federated execution has `x` throw an error. This turns all of `data` into `null` (not `{ x: null }`). But then this re-execution adds another error saying that the non-null field is null.

Take a look at https://codesandbox.io/s/angry-raman-uuzuf?file=/src/index.js for an example of what's going on here.
